scio-core/src/main/scala/com/spotify/scio/estimators/ApproxDistinctCounter.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2020 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.estimators
+
+import com.spotify.scio.coders.{Coder, CoderMaterializer}
+import com.spotify.scio.util.TupleFunctions._
+import com.spotify.scio.values.SCollection
+import org.apache.beam.sdk.{transforms => beam}
+
+/**
+ * Approximate distinct element counter for type `T`, e.g. HyperLogLog or HyperLogLog++. This has two APIs one
+ * estimate total distinct count for a given SCollection and second one estimate distinct count per each key in a
+ * key-value SCollection.
+ *
+ * @tparam T
+ */
+trait ApproxDistinctCounter[T] {
+
+  /**
+   * Return a SCollection with single (Long)value which is the estimated distinct count in the given SCollection with
+   * type `T`
+   */
+  def estimateDistinctCount(in: SCollection[T]): SCollection[Long]
+
+  /**
+   * Approximate distinct element per each key in the given key value SCollection.
+   * This will output estimated distinct count per each unique key.
+   */
+  def estimateDistinctCountPerKey[K](
+    in: SCollection[(K, T)]
+  )(implicit koder: Coder[K], voder: Coder[T]): SCollection[(K, Long)]
```

Comment:
We shouldn't need implicit coders here as they can be infered from `in` for 0.10.x

Reply:
tried to build the coder like this `Coder.raw(CoderMaterializer.kvCoder[K, java.lang.Long](in.context))` but it complains `Cannot find an implicit Coder instance for type:   >> K` , is it because this is outside the SColleciton?

Reply:
You have to "localize" the implicits like this: https://github.com/spotify/scio/blob/master/scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala#L50